middle_end/flambda/simplify/simplify_common.ml

```diff
@@ -50,6 +50,49 @@ type simplify_toplevel =
   -> exn_cont_scope:Scope.t
   -> Expr.t * Upwards_acc.t
 
+let is_self_call dacc apply =
+  let denv = DA.denv dacc in
+  match DE.closure_info denv with
+  | Not_in_a_closure -> false
+  | Closure { code_id = fun_code_id;
+              return_continuation = _; exn_continuation = _; } ->
+    begin match Apply.call_kind apply with
+    | Function (Direct { code_id = apply_code_id; _ }) ->
+      Code_id.equal fun_code_id apply_code_id
+    | Method _ | C_call _
+    | Function (Indirect_known_arity _ | Indirect_unknown_arity)
+      -> false
+    end
+  (* CR: This case should not happend, but it's relatively safe to
+     return false here *)
+  | In_a_set_of_closure_but_not_yet_in_a_specific_closure -> false
```

Comment:
Can we remove this CR comment and add one which just says this is safe, even though it shouldn't happen?  Why would it not be safe?

Reply:
It's safe indeed (it should never be unsafe to consider a tail-call as a regular not-tail call), I changed the CR accordingly.